Рефакторинг системы риал-тайм нотификаций ui

bogachev-pa opened the pull request.

bogachev-pa committed:
use List instead of plain array for NotificationManager (bogachev-pa/rdo-xtext#30)

bogachev-pa committed:
unify ui real-time updating scheme (bogachev-pa/rdo-xtext#30)

bogachev-pa committed:
use List instead of ArrayList for declarations

bogachev-pa committed:
wrapped updater configurations into separate methods (bogachev-pa/rdo-xtext#30)

bogachev-pa committed:
Merge branch 'release/current' of github.com:aurusov/rdo-xtext into feature/notification_rework

bogachev-pa committed:
remove Notifier interface (bogachev-pa/rdo-xtext#30)

it is implemented by one class and no other future uses are seen

bogachev-pa committed:
use enums instead of strings for notification categories (bogachev-pa/rdo-xtext#30)

LeKaitoW commented on ru.bmstu.rk9.rao.lib/src/ru/bmstu/rk9/rao/lib/notification/NotificationManager.java:
```diff
@@ -1,29 +1,42 @@
-		for (Subscriber s : subscribtions.get(category).subscribers)
-			s.fireChange();
+	private final Map<E, Subscription> sub;
+
```
`sub` -> `subscriptions`?

LeKaitoW commented on ru.bmstu.rk9.rao.lib/src/ru/bmstu/rk9/rao/lib/notification/NotificationManager.java:
```diff
@@ -1,29 +1,42 @@
+	public NotificationManager(Class<E> enumClass) {
+		sub = new EnumMap<E, Subscription>(enumClass);
+
+		for (E s : enumClass.getEnumConstants())
```
`s` -> `category`???

LeKaitoW commented on ru.bmstu.rk9.rao.ui/src/ru/bmstu/rk9/rao/ui/run/ExecutionHandler.java:
```diff
@@ -85,6 +85,60 @@ public void run() {
+	private final void configureRealTimeSubscriptions(long startTime) {
+		final Display display = PlatformUI.getWorkbench().getDisplay();
+
+		RealTimeUpdater.addScheduledAction(new Runnable() {
```
Может быть стоит перенести `Runnable` в `StatusView`?

LeKaitoW commented on ru.bmstu.rk9.rao.ui/src/ru/bmstu/rk9/rao/ui/run/ExecutionHandler.java:
```diff
@@ -111,9 +165,6 @@ public Object execute(ExecutionEvent event) throws ExecutionException {
 		final Job run = new Job(project.getName() + " execution") {
 			protected IStatus run(IProgressMonitor monitor) {
 				URLClassLoader cl = null;
```
`cl` -> `classLoader`?

LeKaitoW commented on ru.bmstu.rk9.rao.ui/src/ru/bmstu/rk9/rao/ui/simulation/SimulationModeDispatcher.java:
```diff
@@ -15,13 +16,11 @@ public static void setMode(String currentState) {
 			switch (currentState) {
 			case "P":
 			case "NA":
```
`P` -> `Pause`?
`NA` -> `NoAnimation`?
Может, описывать состояние enum, а не строкой?

LeKaitoW commented on ru.bmstu.rk9.rao.ui/src/ru/bmstu/rk9/rao/ui/simulation/SimulationModeDispatcher.java:
```diff
@@ -15,13 +16,11 @@ public static void setMode(String currentState) {
+				RealTimeUpdater.setPaused(true);
 				break;
 			case "FF":
 			case "NS":
```
Аналогично.

bogachev-pa committed:
rename notification manager fields/variables

bogachev-pa committed:
move StatusView Runnable from ExecutionHandler to StatusView

bogachev-pa committed:
rename ExecutionHandler fields/variables

aurusov commented on ru.bmstu.rk9.rao.lib/src/ru/bmstu/rk9/rao/lib/notification/NotificationManager.java:
```diff
@@ -1,29 +1,42 @@
 
-public class NotificationManager implements Notifier {
-	protected Map<String, Notifier.Subscription> subscribtions;
+public class NotificationManager<E extends Enum<E>> {
```
`E` -> `Category`

bogachev-pa committed:
use enums instead of strings for values passed from plugin extensions

aurusov commented on ru.bmstu.rk9.rao.lib/src/ru/bmstu/rk9/rao/lib/notification/NotificationManager.java:
```diff
@@ -1,29 +1,42 @@
+			subscribers = new LinkedList<Subscriber>();
+		}
+
+		public Subscription addSubscriber(Subscriber object) {
```
`object` -> `subscriber`

aurusov commented on ru.bmstu.rk9.rao.lib/src/ru/bmstu/rk9/rao/lib/notification/NotificationManager.java:
```diff
@@ -1,29 +1,42 @@
+			return this;
+		}
+
+		public boolean removeSubscriber(Subscriber object) {
```
`object` -> `subscriber`

aurusov replied:
Вообще не нашел вызова этого метода, по крайней мере в этом пул-реквесте, может где и есть. Но если нет, то это офигительный залёт в проектировании. Вот для таких косячных решений и нужен гребанный сборщик мусора. Чуть ниже сказано как с этим быть, вот тут https://github.com/aurusov/rdo-xtext/pull/88/files#r34421534.

bogachev-pa replied:
Вызова метода и нет. Удалил его.

aurusov replied:
НЕТ!!! Читай связанные комменты.

aurusov replied:
> Но если нет, то это офигительный залёт в проектировании. 

Объяснял же это голосом, все покивали, хорошо мол. А теперь `Удалил его.` Я в шоке. Еще раз объяснить ?

aurusov commented on ru.bmstu.rk9.rao.lib/src/ru/bmstu/rk9/rao/lib/notification/NotificationManager.java:
```diff
@@ -1,29 +1,42 @@
+
+		public Subscription addSubscriber(Subscriber object) {
+			subscribers.add(object);
+			return this;
```
~~Важно вернуть себя ?~~

Увидел где используется, пусть останется.

bogachev-pa replied:
Это для билдер-стайл добавления субскрайберов:

``` java
subscribtion.addSubcriber("foo").addSubscriber("bar");
```

aurusov replied:
Вообще спорный момент, но давай с этим поживём.

aurusov commented on ru.bmstu.rk9.rao.lib/src/ru/bmstu/rk9/rao/lib/notification/NotificationManager.java:
```diff
@@ -1,29 +1,42 @@
+		}
+
+		public boolean removeSubscriber(Subscriber object) {
+			return subscribers.remove(object);
```
Кто-то обрабатывает результат ?

bogachev-pa committed:
RealTimeUpdater: make update period lower

aurusov commented on ru.bmstu.rk9.rao.ui/src/ru/bmstu/rk9/rao/ui/animation/AnimationView.java:
```diff
@@ -120,46 +117,22 @@ public static void deinitialize() {
+				frameView.redraw();
+			}
 		}
 	};
```
А почему бы не подписаться сразу после создания ?

bogachev-pa replied:
Формально из этого класса мы не можем знать, существует ли уже симулятор или нет. Поэтому решает, когда подписывать тот, кто знает (`ExecutionHandler`). Я бы с удовольствием пораспиливал его, но пока не представляю, как.

aurusov replied:
Подписываться надо по схеме, описанной выше.

aurusov commented on ru.bmstu.rk9.rao.ui/src/ru/bmstu/rk9/rao/ui/notification/RealTimeUpdater.java:
```diff
@@ -0,0 +1,75 @@
+			timerTask.run();
+	}
+
+	private static Display display;
```
Все методы и атрибуты класса статические. Это удобно ? Может быть создать от класса объект, который будет доступен через какой-нибудь класса симулятора ? Мне кажется, что вызывать `start()` до создания базы `Simulator.getDatabase()` просто нельзя. Сейчас это неочевидно, если смотреть только на интерфейс класса.

bogachev-pa replied:
Думаю, стоит поместить его в симулятор, а метод `start()` вызывать ему из конструктора `Database`.

bogachev-pa replied:
Но есть один серьезный недостаток в этом подходе: `RealTimeUpdater` использует `display.asyncExec()` для обновления его компонентов. Формально мы может использовать это в либе, но с точки зрения дизайна это неправильно, потому что это исключительно уишный вызов.

Тут два варианта: либо оставить класс в `ui`, либо вызывать `asyncExec()` руками в каждом субскрайбере.

aurusov replied:
Во, ура. Оказывается для ui есть тоже базовые классы. И их надо научить взаимодействовать с невизуальным ядром. Звучит нормально. Как вариант - из ядра классы по имени, http://stackoverflow.com/a/6094602

bogachev-pa replied:
Про вызов по имени не понял. Класс `Display` мы и так можем поимпортить в либе, это `swt`шный класс, никаких дополнительных зависимостей нам прописывать не надо.

Но мне не нравится идея того, что ядро вообще что-то знает о существовании `ui`. Его вполне может и не быть. Ядра это не касается.

aurusov replied:
> Но мне не нравится идея того, что ядро вообще что-то знает о существовании `ui`.

Именно поэтому ядро получит String, значение которго ему не ведомо !!! Про ЗНАЧЕНИЕ будет знать только вызывающая сторона, а ядро просто интегрирует созданных объект в себя. Могу предложить еще решения, но только после того, как осознаешь это.

bogachev-pa replied:
Я это осознал. Получится неотлаживаемый ад.

aurusov replied:
Это **другой** вопрос. Про читаемость ничего не говорил. Я говорил про разрыв связей между ядром и ui. Связь так порвется ?

bogachev-pa replied:
Не уверен, что вы имеете в виду под словом "порвется", но ядро не будет ничего знать про `ui`, и это ок. "Не понял" относилось скорее к тому, зачем вообще так делать.

В качестве более адекватной альтернативы могу предложить использовать некий `Runnable`, который внутри себя будет содержать другой `Runnable` и в своем методе `run()` будет вызывать `display.asyncExec(innerRunnable)`.

aurusov replied:
По сути да. Но лучше не злоупотребление абстракцией `Runnable`. Лучше вводить конктерные интерфейсы. Ну а ядро будет подписчиком для них. Т.е. связь будет через известный ядру интерфейс типа

``` java
public interface ScheduledAction {
    void doScheduledAction();
}
```

aurusov commented on ru.bmstu.rk9.rao.ui/src/ru/bmstu/rk9/rao/ui/notification/RealTimeUpdater.java:
```diff
@@ -0,0 +1,75 @@
+		timer.scheduleAtFixedRate(timerTask, delay, period);
+	}
+
+	public final static void cancel() {
```
`cancel()` -> `stop()`

bogachev-pa replied:
Называл по аналогии с методами таймера. В принципе, можно и `stop()`.

aurusov replied:
Пока есть `start()`, то другого нельзя.

aurusov commented on ru.bmstu.rk9.rao.ui/src/ru/bmstu/rk9/rao/ui/run/ExecutionHandler.java:
```diff
@@ -83,7 +83,46 @@ public void run() {
+				.addSubscriber(SerializedObjectsView.commonUpdater);
+
+		SerializedObjectsView.commonUpdater.fireChange();
+		TraceView.commonUpdater.fireChange();
```
- Вот чтобы такое не писать, объект должен получать данные сразу после подписки.
- Очень странно, что `ExecutionHandler` хоть что-то знает про эти два класса. Кажется, что подписку надо перенести в них.

bogachev-pa replied:
1. Ок, сделаю.
2. Это делает `ExecutionHandler`, т.к. иначе не понятно, в какой момент им нужно подписываться.

bogachev-pa replied:
1. С другой стороны, это неправильно. Например, мы подписались на окончание прогона. Это же не значит, что нам нужно сразу вызвать `fireChange()` - прогон только начался. Тут другая ошибка в дизайне - вместо ручного `fireChange()`, надо подписать их на начало симуляции.

Edit: done.

aurusov replied:
1. Как вариант
2. Я же объяснил как определить когда. В моей схеме есть место упрощенному `ExecutionHandler`'у

aurusov commented on ru.bmstu.rk9.rao.ui/src/ru/bmstu/rk9/rao/ui/run/ExecutionHandler.java:
```diff
@@ -83,7 +83,46 @@ public void run() {
+				.addScheduledAction(SerializedObjectsView.realTimeUpdateRunnable);
+		RealTimeUpdater
+				.addScheduledAction(AnimationView.realTimeUpdateRunnable);
+	}
```
Опять получился очень много-знающий объект-менеджер всего сразу.

bogachev-pa replied:
Он таким всегда и был. См. выше. Надо координально менять дизайн, чтобы избавиться от вездесущность `ExecutionHandler`'а.

aurusov replied:
Я же написал это было. Но менять НАДО. Сейчас полный отстой. `Я не буду исправлять ошибку, потому что она была раньше` - ты это хочешь сказать ?

aurusov commented on ru.bmstu.rk9.rao.ui/src/ru/bmstu/rk9/rao/ui/run/ExecutionHandler.java:
```diff
@@ -187,106 +224,40 @@ protected IStatus run(IProgressMonitor monitor) {
-
-					ConsoleView
-							.addLine("Started model " + project.getName());
+					configureCommonSubscriptions();
```
Чтобы избавиться от объекта-менеджера, можно вот тут выбросить системное уведомление о запуске модели, поймав которое вьюхи сами подпишут себя на все остальное.

PS: Таких точек в этом большом методе может быть несколько.

aurusov replied:
То, что ты вынес это в отдельный метод, то это ok, но сразу хочется чистить дальше.

aurusov commented on ru.bmstu.rk9.rao.ui/src/ru/bmstu/rk9/rao/ui/run/ExecutionHandler.java:
```diff
@@ -305,31 +276,24 @@ public void run() {
-					animationUpdater.cancel();
-
-					cl.close();
+					RealTimeUpdater.cancel();
```
Вот перед этим место стоит послать сообщение конца моделирования, чтобы подписчики смогли отписаться сами.

bogachev-pa replied:
Подписчик имеет всего одну категорию подписки. Поэтому их отписывание контролируется `RealTImeUpdater`'ом.

aurusov replied:
Подписка/отписка - симметричные операции. Должны контролироваться одним и тем же объектом. В этом и есть ошибка.

bogachev-pa replied:
Т.е. так: `ExecutionHandler` (или `Simulator` статически) посылает всем уведомления, что симулятор проинициализирован, они подписываются, потом посылаем уведомление, что все закончилось, они отписываются? Но чтобы им разослать такое уведомление, надо чтобы они уже были на что-то подписаны. На какую-то из категорий симулятора они подписаться не могут, потому что его `notificationManager` нестатический и может быть еще не создан. Тогда нам нужен еще и какой-то статический `notificationManager`?

aurusov replied:
Да, именно такая схема - каскадные подписки. Надо подписываться на то, на что можно подписаться здесь и сейчас, а не на объект в будущем (для таких сложных схем надо заводить отдельную проксю, как в случае подписок по таймеру).

Статиков лучше избегать. Если их делать, то в минимальном количестве. Относится к коду в целом.

Что тебе мешает пробрасывать симулятор в конструкторы этих объектов ? Внутри конструкторов они смогут подписываться на запуск/остановку модели. А уже внутри этих обработчиков будет симметричная работа со всеми другими подписками.

На выходе мы получим проблему - невозможность отписаться от первого каскада подписок. Но сейчас еще хуже - мы не можем отписаться ни от кого.

Проблему первого каскада можно решить через статического подписчика, который будет рассылть уведомления типа `SIMULATOR_CRAETED`, `SIMULATOR_DESTROYED`. Но как вызвать последнюю - непонятно, надо копать Эклипс. Т.е. проблема первого каскада остаёются, но явно сужается.

bogachev-pa replied:
Конструкторы каких объектов? Вьюх? Субскрайберов, которые живут внутри вьюх? Т.е. субскрайберов создавать им не сразу, а в какой-то момент времени?

Я вас совсем не понимаю, давайте обсудим голосом сегодня вечером?

aurusov replied:
> Вьюх?

Да.

> Т.е. субскрайберов создавать им не сразу, а в какой-то момент времени?

Да. Точнее первый каскад - в конструкторе, второй - как придет уведомление по первому каскаду. Третий каскад создается по уведомлению из второго. И т.д.

> давайте обсудим голосом сегодня вечером?

Готов днем.

aurusov replied:
Обновил ответ выше.

aurusov commented on ru.bmstu.rk9.rao.ui/src/ru/bmstu/rk9/rao/ui/serialization/SerializedObjectsView.java:
```diff
@@ -184,40 +183,17 @@ public final static boolean readyForInput() {
-		public void fireChange() {
-			haveNewRealTimeData = true;
+		public void run() {
+			updateAllOpenedCharts();
```
Зачем это тут ? Почему бы не подписать графики отдельно от этого объекта ?

bogachev-pa replied:
Собираюсь так и сделать.

aurusov commented on ru.bmstu.rk9.rao.ui/src/ru/bmstu/rk9/rao/ui/simulation/SimulationSynchronizer.java:
```diff
@@ -10,7 +10,26 @@
+		}
+
+		public static final ExecutionMode get(final String type) {
+			for (final ExecutionMode t : values()) {
```
Ну что за `t` ?
- `value` или
- `executionMode`

aurusov commented on ru.bmstu.rk9.rao.ui/src/ru/bmstu/rk9/rao/ui/simulation/SimulationSynchronizer.java:
```diff
@@ -10,7 +10,26 @@
+
+		public String getString() {
+			return type;
+		}
```
Почему `getString()` ?

``` java
@Override
public String toString() {
    return type;
}
```

bogachev-pa replied:
Не хотелось оверрайдить такой общий метод, чтобы не путаться в смысле.

aurusov replied:
Хмм. Это и есть java-way. Не понимаю, почему метод, представляющий объект в виде строчки, должен называться в джабе по другому.

aurusov commented on ru.bmstu.rk9.rao.ui/src/ru/bmstu/rk9/rao/ui/simulation/SimulationSynchronizer.java:
```diff
@@ -10,7 +10,26 @@
+			}
+			throw new SimulationSynchronizerException(
+					"Unknown simulation mode: " + type);
+		}
```
Может назвать `getByString()`, чтобы не хотелось его часто вызывать ?

aurusov commented on ru.bmstu.rk9.rao.ui/src/ru/bmstu/rk9/rao/ui/trace/ExportTraceHandler.java:
```diff
@@ -21,18 +21,40 @@
+
+		public String getString() {
+			return type;
+		}
```
Те же три замечания, как по энуму выше

bogachev-pa committed:
code clean up

variables/fields/methods renaming

bogachev-pa committed:
synchronize simulator initialization, add EXECUTION_STARTED notification

bogachev-pa committed:
rename methods/fields/variables in notification system

bogachev-pa committed:
complete notification system refactoring (bogachev-pa/rdo-xtext#30)

- implement two-level subscribtion scheme
- let views and runtime ui components control subscriptions themselves
- change notifier api

bogachev-pa committed:
Merge branch 'release/current' of github.com:aurusov/rdo-xtext into refactoring/notification_rework

bogachev-pa committed:
allow to slow down simulation during dpt seqrch (bogachev-pa/rdo-xtext#30)

bogachev-pa committed:
integrate graph visualization into common subscription scheme

related issue: bogachev-pa/rdo-xtext#30

bogachev-pa committed:
add runtime components deinitialization (bogachev-pa/rdo-xtext#30)

bogachev-pa committed:
fixed order of initialization statements in RuntimeComponents

bogachev-pa committed:
make plot views have their own database parser each

related issue: bogachev-pa/rdo-xtext#30

bogachev-pa committed:
fix SubscriberRegistrationManager initialization (bogachev-pa/rdo-xtext#30)

No notifications about old initialiizations/deinitializations should
be made.

bogachev-pa committed:
make graph view have their own database parser each

related issue: bogachev-pa/rdo-xtext#30

bogachev-pa committed:
slight GraphView methods reordering for better code readability

bogachev-pa committed:
removed unused code

bogachev-pa committed:
remove reductant notifyError() method from Simulator

bogachev-pa committed:
fix watching over simulation aborted state (bogachev-pa/rdo-xtext#30)

aurusov commented on ru.bmstu.rk9.rao.lib/src/ru/bmstu/rk9/rao/lib/dpt/DecisionPointSearch.java:
```diff
@@ -110,11 +111,19 @@ public int compare(GraphNode x, GraphNode y) {
-		this.allowSearch = false;
+		Simulator.getExecutionStateNotifier().addSubscriber(executionAbortedListener,
+				Simulator.ExecutionState.EXECUTION_ABORTED);
 	}
```
Формально можно обойтись без переменной `executionAbortedListener`. Мне она усложнила читаемость, т.к. подумал, что она где-то еще используется и пришлось проверить по коду. Хочется попоробовать без неё. Но на самом деле я не знаю будет ли лучше читаться.

PS: Хотя нет, она нужна, чтобы не создавать подписчика каждый раз заново. Давайте оставим.

aurusov commented on ru.bmstu.rk9.rao.lib/src/ru/bmstu/rk9/rao/lib/dpt/DecisionPointSearch.java:
```diff
@@ -110,11 +111,19 @@ public int compare(GraphNode x, GraphNode y) {
+		}
+	};
+
 	private volatile boolean allowSearch = true;
```
``` java
private volatile boolean allowSearch = false;
```

aurusov commented on ru.bmstu.rk9.rao.lib/src/ru/bmstu/rk9/rao/lib/dpt/DecisionPointSearch.java:
```diff
@@ -110,11 +111,19 @@ public int compare(GraphNode x, GraphNode y) {
 
 	@Override
 	public void fireChange() {
-		this.allowSearch = false;
```
``` java
allowSearch = true;
```

И вообще, предлагаю перенести все тело этого метода в отдельный метод

``` java
private final void onSimulatorInitialized()
```

bogachev-pa replied:
Добавил в `DecisionPointSearch` отдельного субскрайбера, а сама она теперь этот интерфейс не имплементит. Ну и по имени субскрайбера понятно, что это `onSimulatorInitialized`.

aurusov commented on ru.bmstu.rk9.rao.lib/src/ru/bmstu/rk9/rao/lib/notification/Notifier.java:
```diff
@@ -1,26 +1,40 @@
-		}
+	public Notifier(Class<Category> enumClass) {
+		subscriptions = new ConcurrentHashMap<Category, Subscription>();
+		hadNotifications = new ConcurrentHashMap<Category, Boolean>();
```
Кажется правильным хранить мапу на пару
`Category -> pair<Subscription, boolean>`
вместо двух мап.

bogachev-pa replied:
Done.

aurusov commented on ru.bmstu.rk9.rao.lib/src/ru/bmstu/rk9/rao/lib/notification/Subscription.java:
```diff
@@ -0,0 +1,19 @@
+			.newSetFromMap(new ConcurrentHashMap<Subscriber, Boolean>());
+
+	Subscription addSubscriber(Subscriber subscriber) {
+		subscribers.add(subscriber);
```
Думаю, надо проверить результат на `false` и бросить исключение о повторном подключении.

bogachev-pa replied:
Done.

aurusov commented on ru.bmstu.rk9.rao.lib/src/ru/bmstu/rk9/rao/lib/notification/Subscription.java:
```diff
@@ -0,0 +1,19 @@
+	}
+
+	void removeSubscriber(Subscriber subscriber) {
+		subscribers.remove(subscriber);
```
Тут тоже самое, только текст исключения другой.

bogachev-pa replied:
Done.

aurusov commented on ru.bmstu.rk9.rao.lib/src/ru/bmstu/rk9/rao/lib/notification/Notifier.java:
```diff
@@ -1,26 +1,40 @@
 
-	public String[] getAvailableSubscriptions();
+	public void addSubscriber(Subscriber subscriber, Category category,
+			boolean notifeAboutOldData) {
```
`notifeAboutOldData` -> `notifeAboutAccumulatedData`

bogachev-pa replied:
Done

aurusov commented on ru.bmstu.rk9.rao.lib/src/ru/bmstu/rk9/rao/lib/simulator/Simulator.java:
```diff
@@ -10,47 +10,54 @@
-	public static boolean isInitialized() {
-		return INSTANCE != null;
+		simulatorState = SimulatorState.INITIALIZED;
+		simulatorStateNotifier.notifySubscribers(SimulatorState.INITIALIZED);
```
`simulatorStateNotifier.notifySubscribers(simulatorState);`

Давай заведем метод `setSimulatorState` внутри которого будет и измение атрибута и отправка уведомления.

bogachev-pa replied:
Done.

aurusov commented on ru.bmstu.rk9.rao.lib/src/ru/bmstu/rk9/rao/lib/simulator/Simulator.java:
```diff
@@ -10,47 +10,54 @@
-	public static boolean isRunning() {
-		return isRunning;
-	}
+	private static SimulatorState simulatorState = SimulatorState.DEINITIALIZED;
```
Формально тут можно сделать UNDEFINED, ведь чтобы стать DEINITIALIZED надо сначала стать INITIALIZED. Но т.к. непонятны бонусы, то пока так оставим.

aurusov commented on ru.bmstu.rk9.rao.lib/src/ru/bmstu/rk9/rao/lib/simulator/Simulator.java:
```diff
@@ -10,47 +10,54 @@
-	public static synchronized void initSimulation() {
+	public static synchronized void initSimulation(JSONObject modelStructure) {
 		if (isRunning)
 			return;
```
А почему бы тут не выбросить исключение ?

bogachev-pa replied:
Done.
Вообще, мне не нравится наша текущая ситуация с эксепшенами. По каждому чиху мы создаем новый класс, я думаю, это неправильно. Не лучше ли иметь один класс `SimulationException` на всю либу? Для гуи мб имеет смысл иметь раздельные.

aurusov replied:
Я тоже не понимаю зачем плодить классы. Их может быть 2-3 на всю программу.

aurusov commented on ru.bmstu.rk9.rao.lib/src/ru/bmstu/rk9/rao/lib/simulator/Simulator.java:
```diff
@@ -137,16 +132,16 @@ private boolean checkTerminate() {
 	private volatile boolean executionAborted = false;
```
Что можно сказать про
- `isRunning == false` и `executionAborted == true` - это разные состояния ?
- `isRunning == true` и `executionAborted == true` - такое вообще возможно ?

В этих вопросах намек, что стоит подумать над объединением двух булов с 4-мя состояниями в один емун с тремя. Но в другом реквесте. Сначада давай тут или в отдельной таске обсудим.

bogachev-pa replied:
Как минимум одно из них статическое, а другое нет. Я пока не очень представляю, как разделить.

aurusov replied:
Но ведь есть пример, который не возможен с точки зрения логики. Или у нас возможен ?

bogachev-pa replied:
Например, пользователь прервал симуляцию кнопкой `stop` - у симулятора статически выставился статус `simulationAborted = true`, но мы сейчас (в другом потоке!) находимся в процессе проверки дпт и сможем остановить симуляцию по-настоящему только когда выйдем из него. Тогда-то и выставится `isRunning = false`.

aurusov replied:
Да, выглядят как разные состояния. Давай так и оставим.

aurusov commented on ru.bmstu.rk9.rao.ui/src/ru/bmstu/rk9/rao/ui/RaoActivatorExtension.java:
```diff
@@ -39,12 +40,7 @@
 	public void start(BundleContext context) throws Exception {
 		super.start(context);
```
Метод запускается только один раз на плагин ? Т.е. не при каждом прогоне модели ?

bogachev-pa replied:
Один раз при запуске плагина, если я правильно понимаю.

aurusov commented on ru.bmstu.rk9.rao.ui/src/ru/bmstu/rk9/rao/ui/animation/AnimationView.java:
```diff
@@ -107,6 +107,10 @@ public static void initialize(ArrayList<AnimationFrame> frames) {
+
+		ExecutionMode currentMode = SimulationModeDispatcher.getMode();
+		disableAnimation(currentMode == ExecutionMode.NO_ANIMATION ? true
+				: false);
```
Метод с отрицанием в имени напрягает. Давай инвертируем логику и начнем с параметра `noAnimation` -> `enableAnimation`. Тогда `disableAnimation()` -> `setEnableAnimation()` и т.д.

bogachev-pa replied:
Done.

aurusov commented on ru.bmstu.rk9.rao.ui/src/ru/bmstu/rk9/rao/ui/animation/AnimationView.java:
```diff
@@ -286,4 +270,12 @@ private static boolean canDraw() {
 				&& currentFrame != null;
 	}
+
+	private final SubscriberRegistrationManager subscriberRegistrationManager = new SubscriberRegistrationManager();
```
Что-то со стилем.

bogachev-pa replied:
Глупый эклипсовский форматер не умеет переносить такие строчки.

aurusov commented on ru.bmstu.rk9.rao.ui/src/ru/bmstu/rk9/rao/ui/animation/AnimationView.java:
```diff
@@ -96,7 +96,7 @@ private static void initializeFrames() {
 
-	public static void initialize(ArrayList<AnimationFrame> frames) {
+	public static void initialize(java.util.List<AnimationFrame> frames) {
 		isRunning = true;
```
`isRunning` -> `isInitialized`

bogachev-pa replied:
Done

aurusov commented on ru.bmstu.rk9.rao.ui/src/ru/bmstu/rk9/rao/ui/animation/AnimationView.java:
```diff
@@ -269,6 +249,10 @@ public void createPartControl(Composite parent) {
 			initializeFrames();
+
+		subscriberRegistrationManager
+				.enlistRealTimeSubscriber(realTimeUpdateRunnable);
```
`enlistRealTimeSubscriber` говорит том, что это объект - прокся. Мне кажется, что это необязательно и будет проще использовать `addSubscriber`. Или у тебя были причины ?

bogachev-pa replied:
Это не прокся, а небольшой враппер над группой методов, которые не хочется писать в каждом классе заново. Поэтому и `enlist`, что мы этим методом никуда никого не регистрируем, а просто перечислям во враппере объекты, с которыми будем работать.

aurusov replied:
С какой целью мы их там перечисляем ?

aurusov commented on ru.bmstu.rk9.rao.ui/src/ru/bmstu/rk9/rao/ui/graph/GraphControl.java:
```diff
@@ -5,21 +5,14 @@
 
-	final static Rectangle monitorBounds = PlatformUI.getWorkbench()
-			.getDisplay().getBounds();
+	private static Rectangle monitorBounds;
```
Можно грохнуть, использовать только как локальную переменную ниже.

bogachev-pa replied:
Done

aurusov commented on ru.bmstu.rk9.rao.ui/src/ru/bmstu/rk9/rao/ui/notification/RealTimeUpdater.java:
```diff
@@ -0,0 +1,115 @@
+	}
+
+	public final void deinitialize() {
+		subscriberRegistrationManager.deinitialize();
```
А почему нет отписывания от `EXECUTION_STARTED/EXECUTION_COMPLETED` ?

bogachev-pa replied:
Вся двухуровневая подписка/отписка реализуется внутри `SubscriberRegistrationManager`.

aurusov replied:
Там реализуется позднее связывание/развязывание. А кто быдет вызывать `subscribersInfo.remove`/`realTimeSubscribersInfo.remove` ?

aurusov commented on ru.bmstu.rk9.rao.ui/src/ru/bmstu/rk9/rao/ui/notification/RealTimeUpdater.java:
```diff
@@ -0,0 +1,115 @@
+			}
+		};
+		timer.scheduleAtFixedRate(timerTask, delay, period);
+		this.paused = (SimulationModeDispatcher.getMode() == ExecutionMode.PAUSE);
```
зачем `this` ?

bogachev-pa replied:
Убрал.

aurusov commented on ru.bmstu.rk9.rao.ui/src/ru/bmstu/rk9/rao/ui/notification/RealTimeUpdater.java:
```diff
@@ -0,0 +1,115 @@
+	private TimerTask timerTask = null;
+
+	private final long delay = 0;
+	private final long period = 40;
```
``` java
private final long delay_msec = 0;
private final long period_msec = 1000/25;
```

bogachev-pa replied:
Done.

aurusov commented on ru.bmstu.rk9.rao.ui/src/ru/bmstu/rk9/rao/ui/plot/PlotDataParser.java:
```diff
@@ -70,42 +58,37 @@ public static void removeIndexFromMaps(Index index) {
+
+		if (lastPatternCount == 0) {
+			dataset.add(new PlotItem(0, lastPatternCount));
 		}
```
А почему `lastPatternCount` ? Предлагаю `patternCount`.

bogachev-pa replied:
Done.

aurusov commented on ru.bmstu.rk9.rao.ui/src/ru/bmstu/rk9/rao/ui/plot/PlotView.java:
```diff
@@ -112,17 +108,82 @@ public void widgetDisposed(DisposeEvent event) {
+	}
+
+	private final void deinitialize() {
+		subscriberRegistrationManager.deinitialize();
```
Не симметричен инициализации.

bogachev-pa replied:
Кажется, оверрайдить `dispose()` тут вообще не нужно, потому что это действие мы все равно выполняем, когда диспозится `plotFrame()`. Ну или наоборот убрать оттуда все это.

aurusov replied:
Я только про подписку.

aurusov commented on ru.bmstu.rk9.rao.ui/src/ru/bmstu/rk9/rao/ui/plot/PlotView.java:
```diff
@@ -112,17 +108,82 @@ public void widgetDisposed(DisposeEvent event) {
 
+	@Override
+	public void dispose() {
+		subscriberRegistrationManager.deinitialize();
```
`deinitialize();`

bogachev-pa replied:
Done.

aurusov commented on ru.bmstu.rk9.rao.ui/src/ru/bmstu/rk9/rao/ui/run/ExecutionHandler.java:
```diff
@@ -175,118 +156,31 @@ protected IStatus run(IProgressMonitor monitor) {
-
-					SimulationModeDispatcher.setMode((String) state.getValue());
-
 					display.syncExec(() -> AnimationView.initialize(frames));
```
Это разве не нужно убрать ? Или возникли проблемы ?

bogachev-pa replied:
Надо убрать. Пока не сделал.

aurusov replied:
В этом пул-реквесте ?

bogachev-pa replied:
Можно и в этом, но займет какое-то время, т.к. нужна нормальная модель с анимацией для отладки.

aurusov replied:
Я хотел бы замержить его как можно быстрее. Продолжить чистку можно в отдельной задаче по анимации.

aurusov commented on ru.bmstu.rk9.rao.ui/src/ru/bmstu/rk9/rao/ui/serialization/SerializationConfigView.java:
```diff
@@ -184,10 +187,30 @@ public void modelChanged(XtextResource resource) {
+		subscriberRegistrationManager.enlistCommonSubscriber(enableSubscriber,
+				ExecutionState.EXECUTION_COMPLETED).enlistCommonSubscriber(
+						disableSubscriber, ExecutionState.EXECUTION_STARTED);
+		subscriberRegistrationManager.initialize();
```
А где отписка ?

bogachev-pa replied:
Добавил.

aurusov commented on ru.bmstu.rk9.rao.ui/src/ru/bmstu/rk9/rao/ui/serialization/SerializedObjectsView.java:
```diff
@@ -136,41 +132,15 @@ public void doubleClick(DoubleClickEvent event) {
+				commonUpdater, ExecutionState.EXECUTION_COMPLETED);
+		subscriberRegistrationManager
+				.enlistRealTimeSubscriber(realTimeUpdateRunnable);
+		subscriberRegistrationManager.initialize();
```
И тут нет отписки. Лучше добавить два метода `initialize()/deinitialize()`.

bogachev-pa replied:
Она есть в методе `dispose()`. Перенес повыше для читаемости. Насчет методов - не хочется их плодить в таком количестве. В `createPartControl()` регистрируем, в `disposer()` разрегистрируем, симметрично же.

aurusov replied:
1. Нужна симметрия
2. В dispose() не всё

aurusov commented on ru.bmstu.rk9.rao.ui/src/ru/bmstu/rk9/rao/ui/serialization/SerializedObjectsView.java:
```diff
@@ -239,11 +184,18 @@ public void run() {
+
+	@Override
+	public void dispose() {
+		subscriberRegistrationManager.deinitialize();
```
Только `deinitialize()`, которого еще нет.

aurusov replied:
ping ?

aurusov commented on ru.bmstu.rk9.rao.ui/src/ru/bmstu/rk9/rao/ui/simulation/SimulationModeDispatcher.java:
```diff
@@ -1,29 +1,46 @@
-			}
+	public static void setMode(ExecutionMode currentMode) {
+		if (!RuntimeComponents.isInitialized())
+			return;
```
Может всетаки исключение ?

bogachev-pa replied:
Старая проверка осталась, такого вообще не может быть. Ну или кидаем исключение (получается, мы используем их как `ASSERT`).

aurusov replied:
Да. Если такого не может быть, то зачем стоит проверка ?

bogachev-pa replied:
Ну я же говорю, осталась старая, просто забыл удалить. А теперь с исключением она играет роль `BUG_ON()` или `ASSERT()`, т.е. для выявление внутренних ошибок в системе.

aurusov replied:
Если логика гарантирует, то проверку лучше убрать.

bogachev-pa replied:
Нотификации достаточно сложная система и лишняя проверка на внутренние ошибки, на мой взгляд, не помешает.

aurusov commented on ru.bmstu.rk9.rao.ui/src/ru/bmstu/rk9/rao/ui/simulation/SimulationSynchronizer.java:
```diff
@@ -7,53 +7,62 @@
 	public static enum ExecutionMode {
-		NO_ANIMATION, FAST_FORWARD, NORMAL_SPEED, PAUSE
-	}
+		PAUSE("P"), NO_ANIMATION("NA"), FAST_FORWARD("FF"), NORMAL_SPEED("NS");
```
Текстовые значения где используются ?
PS: нашел два места. В них нельзя отказаться от строк ?

bogachev-pa replied:
Нет, т.к. пробросом строк занимается эклипс через хэндлеры кнопок/менюшек. В более умные енумы он не умеет. (или умеет, но очень нетривиально)

aurusov commented on ru.bmstu.rk9.rao.ui/src/ru/bmstu/rk9/rao/ui/simulation/SimulationSynchronizer.java:
```diff
@@ -7,53 +7,62 @@
-	public static void finish() {
-		INSTANCE = null;
+	public final void deinitialize() {
+		subscriberRegistrationManager.deinitialize();
```
Нужна симметрия с подпиской.

bogachev-pa replied:
Сделал. Но, может быть, тогда называть эти методы `initialize/deinitializeSubscriptions()`. Тогда можно будет и симметрию везде соблюсти.

aurusov replied:
Да.

aurusov replied:
Симметрии не вижу.

aurusov commented on ru.bmstu.rk9.rao.ui/src/ru/bmstu/rk9/rao/ui/simulation/SimulationSynchronizer.java:
```diff
@@ -78,46 +87,40 @@ public void fireChange() {
 
+	public void setSimulationSpeed(int value) {
 		if (value < 1 || value > 100)
 			return;
```
``` java
if (value < 1 || 100 < value)
```

Может исключение ?

bogachev-pa replied:
Кажется, ноль там все-таки может быть, разве нет? И тогда кидать исключение для `(value < 0 || value >100)`

aurusov replied:
Контрол не позволяет.

aurusov commented on ru.bmstu.rk9.rao.ui/src/ru/bmstu/rk9/rao/ui/simulation/StatusView.java:
```diff
@@ -187,6 +190,10 @@ public void propertyChange(PropertyChangeEvent event) {
+
+		subscriberRegistrationManager
+				.enlistRealTimeSubscriber(realTimeUpdateRunnable);
+		subscriberRegistrationManager.initialize();
```
Нужна отписка.

bogachev-pa replied:
В методе `dispose()`.

aurusov replied:
Симметрия не соблюдена.

aurusov commented on ru.bmstu.rk9.rao.ui/src/ru/bmstu/rk9/rao/ui/simulation/StatusView.java:
```diff
@@ -208,17 +215,40 @@ private void updateScrolledCompositeSize() {
+
+	@Override
+	public void dispose() {
+		subscriberRegistrationManager.deinitialize();
```
`deinitialize();`

bogachev-pa replied:
см. комментарий к https://github.com/aurusov/rdo-xtext/pull/88#discussion_r35490727

aurusov replied:
Не вижу симметрии.

aurusov commented on ru.bmstu.rk9.rao.ui/src/ru/bmstu/rk9/rao/ui/trace/TraceView.java:
```diff
@@ -172,13 +172,12 @@ public void selectionChanged(SelectionChangedEvent event) {
+				commonUpdater, ExecutionState.EXECUTION_COMPLETED);
+		subscriberRegistrationManager
+				.enlistRealTimeSubscriber(realTimeUpdateRunnable);
+		subscriberRegistrationManager.initialize();
```
Нужна отписка, подписку в отдельный метод.

bogachev-pa replied:
Отписка в `dispose()`.

aurusov replied:
Не вижу симметрии.

aurusov commented on ru.bmstu.rk9.rao.ui/src/ru/bmstu/rk9/rao/ui/trace/TraceView.java:
```diff
@@ -422,6 +399,14 @@ public final static boolean readyForInput() {
+
+	@Override
+	public void dispose() {
+		subscriberRegistrationManager.deinitialize();
```
`deinitialize();`

bogachev-pa replied:
см. комментарий к https://github.com/aurusov/rdo-xtext/pull/88#discussion_r35490727

aurusov replied:
Не вижу симметрии.

aurusov commented on ru.bmstu.rk9.rao/src/ru/bmstu/rk9/rao/compilers/ResultCompiler.xtend:
```diff
@@ -229,14 +227,11 @@ class ResultCompiler
-						.addSubscriber(INSTANCE.finalizer);
+					notifier.addSubscriber(INSTANCE, Simulator.ExecutionState.STATE_CHANGED);
+					notifier.addSubscriber(INSTANCE.finalizer, Simulator.ExecutionState.EXECUTION_ABORTED);
+					notifier.addSubscriber(INSTANCE.finalizer, Simulator.ExecutionState.EXECUTION_COMPLETED);
```
А для этого места, и которое чуть ниже, можно придумать отписку ?

bogachev-pa replied:
Думаю, можно, но подписки для не-гуи хотелось бы обсудить голосом.

bogachev-pa committed:
notification system code cleanup (bogachev-pa/rdo-xtext#30)

bogachev-pa committed:
make subscribers registration completely symmetric (bogachev-pa/rdo-xtext#30)

bogachev-pa committed:
simplify internal deffered subscriber manager and its children structure

related issue: bogachev-pa/rdo-xtext#30

bogachev-pa committed:
simplify deffered subscriber manager initialization (bogachev-pa/rdo-xtext#30)

bogachev-pa committed:
fix moment when simulation synchronizer values are set (bogachev-pa/rdo-xtext#30)

bogachev-pa committed:
no more trying to start drawing graph if there are no nodes yet
